File: project2/project2.py
from random import randint

### state

# time, departed count
t = 0
dep = 0
line = []
bathroom = []
# class
class Person:
	def __init__(self,uid):
		# init'd to 0, will change later
		self.leave_time = 0
		# rng for how long they stay
		self.in_time = randint(3,7)
		self.uid = uid
		# weighted rng for gender
		if randint(1,10) >= 6:
			self.gender = 1
		else:
			self.gender = 0

### functions

## utility

# pop next in line
def person_pop():
	p = line.pop(0)
	p.leave_time = t + p.in_time
	return(p)
# pop specified person
def person_remove(d):
	# try catch in case of empty line, should never occur
	try:
		i = line.index(d)
		p = line.pop(i)
		p.leave_time = t + p.in_time
		return(p)
	except:
		print("\n Unexpected happenings")
		print(d.uid)
		print(line)
		print("")
def print_person(uid,g,i):
	print("Time: %2.0f; Person %2.0f (%s) enters the bathroom for %d minutes" % (t, uid, ("M" if g else "F"), i))

## main

# add to bathroom
def Arrive():
	# find last three items on list, or less if it's too small

	global line
	global bathroom
	add_bathroom = []

	if len(bathroom) == 0:
		p = person_pop()
		bathroom.append(p)
		print_person(p.uid + 1, p.gender, p.in_time)
	elif len(bathroom) == 2 and line[0].gender != bathroom[0].gender:
		return

	for f in line:
		index = line.index(f)
		if f.gender == bathroom[0].gender and index < 2 and len(bathroom) < 2:
			p = line[index]
			p.leave_time = t + p.in_time
			add_bathroom.append(line[index])
			print_person(line[index].uid + 1, line[index].gender, line[index].in_time)
		if f.gender == bathroom[0].gender and len(add_bathroom) < 1:
			p = line[index]
			p.leave_time = t + p.in_time
			add_bathroom.append(line[index])
			print_person(line[index].uid + 1, line[index].gender, line[index].in_time)

	line = [f for f in line if f not in add_bathroom]
	bathroom += add_bathroom

# finds those who are done and removes
def UseFacilities():
	# if it's empty
	if not bathroom:
		return
	# if it's not, iterate over every person in there
	for f in bathroom:
		# check if current time is past/equal to their leave time
		if f.leave_time - t <= 0:
			# print(str(f.uid) + " leaves the bathroom.")
			global dep
			dep += 1
			print("Time: %2.0f; Person %2.0f (%s) exits (departure = %2.0f)" % (t, f.uid + 1, ("M" if f.gender else "F"), dep))
			Depart(f)

# removes specified person from bathroom
def Depart(f):
	bathroom.remove(f)

### main

# 3 required cases
case = input("Input case (a,b,c): ")

# 5 at a time
if case.lower() == 'a':
	# count of total people
	c = 0
	# first 5 added
	for f in range(0,5):
		line.append(Person(f))
		print("Time: %2.0f; Person %2.0f (%s) arrives" % (t, f + 1, ("M" if line[f].gender else "F")))
		c += 1

	# central while loop, until count hits 20 and line empties
	while dep < 20:
		# adds 5 more at times
		if t == 10:
			print("\n 5 new arrivals \n")
			for f in range(0,5):
				temp = Person(f + 5)
				line.append(temp)
				print("Time: %2.0f; Person %2.0f (%s) arrives" % (t, temp.uid + 1, ("M" if temp.gender else "F")))
				c += 1
		elif t == 20:	
			print("\n 5 new arrivals \n")
			for f in range(0,5):
				temp = Person(f + 10)
				line.append(temp)
				print("Time: %2.0f; Person %2.0f (%s) arrives" % (t, temp.uid + 1, ("M" if temp.gender else "F")))
				c += 1
		elif t == 30:
			print("\n 5 new arrivals \n")
			for f in range(0,5):
				temp = Person(f + 15)
				line.append(temp)
				print("Time: %2.0f; Person %2.0f (%s) arrives" % (t, temp.uid + 1, ("M" if temp.gender else "F")))
				c += 1

		# remove from bathroom
		UseFacilities()
		# add to bathroom
		if len(bathroom) < 3 and line:
			Arrive()
		# increment time
		t += 1

# 10 at a time
elif case.lower() == 'b':
	# count of total people
	c = 0
	# first 10 added
	for f in range(0,10):
		line.append(Person(f))
		print("Time: %2.0f; Person %2.0f (%s) arrives" % (t, f + 1, ("M" if line[f].gender else "F")))
		c += 1

	# central while loop, until count hits 20 and line empties
	while dep < 20:
		# adds 10 more at times
		if t == 10:
			print("\n 10 new arrivals \n")
			for f in range(0,10):
				temp = Person(f + 10)
				line.append(temp)
				print("Time: %2.0f; Person %2.0f (%s) arrives" % (t, temp.uid + 1, ("M" if temp.gender else "F")))
				c += 1

		# remove from bathroom
		UseFacilities()
		# add to bathroom
		if len(bathroom) < 3 and line:
			Arrive()
		# increment time
		t += 1

# all at once
elif case.lower() == 'c':
	# line generation
	for f in range(0,20):
		line.append(Person(f))
		print("Time: %2.0f; Person %2.0f (%s) arrives" % (t, f + 1, ("M" if line[f].gender else "F")))

	# main while loop
	while dep < 20:
		# remove from bathroom
		UseFacilities()
		# add to bathroom
		if len(bathroom) < 3 and line:
			Arrive()
		# increment time
		t += 1

# improper input
else:
	print("invalid input")

for f in bathroom :
	print(f.uid)

File: project2/project2.c
#include <stdio.h>

void initLine();
int genGen();
void Arrive(int * id, int * gender);
void UseFacilities(int * id, int * gender, int * time);
void Depart(int * id, int * gender);
void OnePerson(int * id, int * gender, int * time);

// 3 people max in restroom at any given time
// 2 people max in restroom if opposite gender is waiting

struct Person {
  int id;
  int gender;
  int timeUsing;
  int timeOut;
  int left;
};

struct Person init[20];		// Where the ID is the place in line
struct Person line[20];
struct Person bathroom[3];
int currentTime = 0;

int main() {
	srand(time(NULL));
	int done = 0;
	for(; done == 0; curentTime++) {
		// Check to see if everyone is done and line is empty
		int checkAllDone = -1;
		for (int i = 0; i < 20; i++) {
			if (init[i].left == 1) {
				checkAllDone += 1;
			}
			else {
				break;
			}
		}
		if (checkAlldone == 19) {
			done = 1;
			break;
		}

		// Check the status of the bathroom
		// Check for empty spot
		int empty[3];
		int totalEmpty = 0;
		for (int i = 0; i < 3; i++) {
			if (bathroom[i] == NULL) {
				empty[i] == 1;
				totalEmpty += 1;
			}
		}
		// Bathroom is empty
		if (totalEmpty == 3) {
			// Line checks
			int currentGender = -1; //No gender yet
			int arrived = 0;		//No one has arrived yet
			int max = 3;			//Current max number of people that can arrive
			for (int i = 0; i < 20; i++) {
				//Make sure next in line is same gender as we are looking for
				if (currentGender == -1 || line[i].gender == currentGender) {
					// Check next in line to see gender
					if (currentGender != -1 && line[i+1].gender != currentGender) {
						// Set max arrived to 2 if next person is opposite gender
						max = 2;
					}
					// Check if arrived is less than max
					if (arrived < max) {
						for (int j = 0; j < 3; j++) {
							// If slot in bathroom empty
							if (bathroom[j] == NULL) {
								// Arrive and leave bathrom loop
								Arrive(line[i]);
								j = 3;
								break;
							}
						}
						// One more person has arrives
						arrived += 1;
					}
				}
				// Check if max has been put into the bathroom
				if (arrived == max) {
					i = 20;
					break;
				}
				
			}
		}

	}

	return 0;
}

//Initialization Functions
void allAtOnce() {
	for (int i = 0; i < 20; i++) {
		line[i].id = i;
		line[i].gender = genGen();	   //Weighted generation function, will write it later
		line[i].timeUsing = rand() % (7 + 1 - 3) + 3;
	}
}

int genGen() {

	int gender = rand() % (100+1);

	if (gender <= 60) {
		return 0;
	}
	else {
		return 1;
	}

}

// Main functions
void Arrive(int * id, int * gender) {
	for (int i = 0; i < 3; i++) {
		if (bathroom[i] == NULL) {
			bathroom[i] == *p;
		}
	}

}

void UseFacilities(int * id, int * gender, int * time) {

}

void Depart(int * id, int * gender) {

}

void OnePerson(int * id, int * gender, int * time) {

}
